crates/iota-data-ingestion-core/src/reducer.rs

```diff
@@ -15,10 +15,40 @@ use crate::{
     worker_pool::WorkerPoolStatus,
 };
 
+/// Processes and commits batches of messages produced by workers.
+///
+/// The Reducer trait provides batch processing capabilities for messages
+/// generated by [`Worker`]s. It allows for custom batching logic and efficient
+/// processing of worker results.
+///
+/// # Batch Processing
+///
+/// [`Messages`](Worker::Message) are accumulated in batches based on the
+/// [`should_close_batch`](Reducer::should_close_batch) policy.
+///
+/// When a batch is
+/// ready to be committed (as determined by `should_close_batch`), the
+/// [`commit`](Reducer::commit) method is called with the collected messages.
 #[async_trait]
 pub trait Reducer<Mapper: Worker>: Send + Sync {
+    /// Commits a batch of messages.
+    ///
+    /// This method is called when a batch is ready to be processed, as
+    /// determined by [`should_close_batch`](Reducer::should_close_batch). The
+    /// implementation should handle the batch processing logic and return
+    /// an error if the operation fails.
```

Comment:
could we remind here that messages passed here are guaranteed to be ordered by checkpoint number?
it may be a useful property for some use cases

Reply:
Checkpoint messages are not inherently ordered upon arrival. The Reducer maintains order by storing messages in a hashmap keyed by checkpoint sequence number. It sequentially builds batches, incrementing the `current_checkpoint_number` with each addition, until `should_close_batch` returns true. This ensures committed batches are strictly ordered by checkpoint number.

logic: https://github.com/iotaledger/iota/blob/develop/crates/iota-data-ingestion-core/src/reducer.rs#L76-L94

Reply:
so after applying the logic that you linked, data passed to the commit() is ordered by checkpoint sequence number, so it may be worth pointing out

or am I missing something?

Reply:
fixed: https://github.com/iotaledger/iota/pull/5735/commits/e12737ff31b033b243aeac8760b2029d98ce07fc